examples/command_bot.rs

```diff
@@ -0,0 +1,99 @@
+use std::ops::Deref;
+use std::sync::Arc;
+use std::{env, process::exit};
+use url::Url;
+
+use matrix_sdk::{
+    self,
+    events::room::message::{MessageEvent, MessageEventContent, TextMessageEventContent},
+    AsyncClient, AsyncClientConfig, EventEmitter, Room, SyncSettings,
+};
+use tokio::sync::Mutex;
+
+struct CommandBot {
+    client: AsyncClient,
+}
+
+impl CommandBot {
+    pub fn new(client: AsyncClient) -> Self {
+        Self { client }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventEmitter for CommandBot {
+    async fn on_room_message(&mut self, room: Arc<Mutex<Room>>, event: Arc<Mutex<MessageEvent>>) {
+        if let MessageEvent {
+            content: MessageEventContent::Text(TextMessageEventContent { body: msg_body, .. }),
+            ..
+        } = event.lock().await.deref()
+        {
+            let room = room.lock().await;
```

Comment:
You are holding on to this lock, the client will try to acquire the same lock when trying to send the message. If you drop the lock before trying to send it it should succeed.

We might want to make it harder to misuse the locking by moving the lock inside the Room struct like I did with all the stuff in `olm.rs`. That way you'll only be able to hold on the lock for a single function call. Though not sure how that will mess with the stuff I'm doing in weechat-matrix.